### DISCUSSION TITLE
Projections over tracked entities

### DISCUSSION THREAD

**Discussion by montyclt** [ravendb/ravendb#15365] [Q&A]
Person class:
```cs
public class Person
{
    public string Id { get; set; }
    public string GivenName { get; set; }
    public string FamilyName { get; set; }
}
```

I'm loading this document from the database and changing the family name ONLY in the session (without saving changes):
```json
{
    "GivenName": "Ivan",
    "FamilyName": "Montilla",
    "@metadata": {
        "@collection": "People",
        "Raven-Clr-Type": "Person, ConsoleApp2"
    }
}
```
```cs
var ivan = await session.LoadAsync("people/1-A");
ivan.FamilyName = "Doe";
```

When retrieving full entities (without projections), the entities tracked by the session are used:
```cs

var ivanFromQuery = await session.Advanced.AsyncRawQuery<Person>("from People where id() = 'people/1-A'").FirsrtAsync();
// At this point, ivan and ivanFromQuery are referenced equals, that means the query taken the tracked entity.
Console.WriteLine(ivanFromQuery.FamilyName); // Doe
```

But if I do a projection, the tracked data are ignored:
```cs
var ivanFromProjection = await session.Advanced.AsyncRawQuery<dynamic>("from People where id() = 'people/1-A' select FamilyName").FirstAsync();
Console.WriteLine(ivanFromProjection.FamilyName); // Montilla
Console.WriteLine(ivan.FamilyName); // Doe
```

I suppose this behaviour is expected and is by design, because the data comes from an index (and will be specially difficult when coming from custom map/reduce index), any idea to do projections that are aware of unsaved changes tracked by the session?

**Comment by ayende** [comment]
Projections are explicitly isolated from the tracked changes.
Consider what happens when you have something like:

```
var ivanFromProjection = await session.Advanced.AsyncRawQuery<dynamic>("from People as p where id() = 'people/1-A' select { Name: p.FirstName, Now: new Date()}").FirstAsync();
```

There is no association between the projection and the document. Changes to the projection aren't tracked, etc.

**Reply by montyclt** [reply]
Possible misunderstood. I'm not talking about tracking projections, but be aware of tracked unsaved entities when doing a projection.

```cs
var ivan = await session.LoadAsync("people/1-A");
ivan.FamilyName = "Doe";
var ivanFromProjection = await session.Advanced.AsyncRawQuery<dynamic>("from People where id() = 'people/1-A' select FamilyName").FirstAsync();
Console.WriteLine(ivanFromProjection.FamilyName); // Montilla. Any option for get "Doe" here? Document people/1-A is tracked, but SaveChanges is not called.
```

**Reply by garayx** [reply]
I guess you can get it from `session.Advanced.WhatChanged()`: https://ravendb.net/docs/article-page/5.1/Csharp/client-api/session/how-to/check-if-there-are-any-changes-on-a-session#whatchanged

**Reply by ayende** [reply]
There is no way to do that. Consider that you may do `load` inside the query, to get additional data from _another_ document.
Basically, the issue is *really* complex and the scenario is simple. If you use entities, we use Unit of Work and you have one instance of a document in the session.
Projections aren't tied to their originating documents, however, and aren't treated in this manner.

**Reply by montyclt** [reply]
RavenDB server has support of transactions? I think that a idea to solve this is start a transaction in the server when a change is tracked and commit it when `SaveChanges` is called.

This is only a idea, I don't know the transactional support of server and how difficult to implement is that. And will introduce a breaking change.

**Reply by ayende** [reply]
Yes, RavenDB has transactions, of course.
But we don't do a transaction per connection, but per request. The idea is that holding a transaction open over TCP socket timeframes is *really* expensive.

**Reply by montyclt** [reply]
Are you talking about maintaining the TCP socket open while operation is in progress? I think other option is that the whole operation generates a transaction identifier that is used across multiple server requests. The server can use this identifier to stack the operations in the same transaction, allowing using different TCP connections over the same transaction.

Again, this is only an idea, and not a request. I don't know the difficulty and considerations of doing that.

**Reply by ayende** [reply]
The socket or HTTP requests is the same. Keeping a transaction open for a long time is really expensive.
Long here is defined as more than a single request/response cycle.

See: https://dl.acm.org/doi/10.1145/1376616.1376713 - 40%+ of DB time was spent managing locks.
